File: src/pose_estimation_book1.cpp
#include "opencv2/opencv.hpp"

int main()
{
    const char *input = "data/blais.mp4", *cover = "data/blais.jpg";
    double f = 1000, cx = 320, cy = 240;
    size_t min_inlier_num = 100;

    // Load the object image and extract features
    cv::Mat obj_image = cv::imread(cover);
    if (obj_image.empty()) return -1;

    cv::Ptr<cv::FeatureDetector> fdetector = cv::ORB::create();
    cv::Ptr<cv::DescriptorMatcher> fmatcher = cv::DescriptorMatcher::create("BruteForce-Hamming");
    std::vector<cv::KeyPoint> obj_keypoint;
    cv::Mat obj_descriptor;
    fdetector->detectAndCompute(obj_image, cv::Mat(), obj_keypoint, obj_descriptor);
    if (obj_keypoint.empty() || obj_descriptor.empty()) return -1;
    fmatcher->add(obj_descriptor);

    // Open a video
    cv::VideoCapture video;
    if (!video.open(input)) return -1;

    // Prepare a box for simple AR
    std::vector<cv::Point3f> box_lower = { cv::Point3f(30, 145,   0), cv::Point3f(30, 200,   0), cv::Point3f(200, 200,   0), cv::Point3f(200, 145,   0) };
    std::vector<cv::Point3f> box_upper = { cv::Point3f(30, 145, -50), cv::Point3f(30, 200, -50), cv::Point3f(200, 200, -50), cv::Point3f(200, 145, -50) };

    // Run pose estimation
    cv::Mat K = (cv::Mat_<double>(3, 3) << f, 0, cx, 0, f, cy, 0, 0, 1);
    cv::Mat dist_coeff = cv::Mat::zeros(5, 1, CV_64F), rvec, tvec;
    while (true)
    {
        // Grab an image from the video
        cv::Mat image;
        video >> image;
        if (image.empty()) break;

        // Extract features and match them to the object features
        std::vector<cv::KeyPoint> img_keypoint;
        cv::Mat img_descriptor;
        fdetector->detectAndCompute(image, cv::Mat(), img_keypoint, img_descriptor);
        if (img_keypoint.empty() || img_descriptor.empty()) continue;
        std::vector<cv::DMatch> match;
        fmatcher->match(img_descriptor, match);
        if (match.size() < min_inlier_num) continue;
        std::vector<cv::Point3f> obj_points;
        std::vector<cv::Point2f> obj_project, img_points;
        for (auto m = match.begin(); m < match.end(); m++)
        {
            obj_points.push_back(cv::Point3f(obj_keypoint[m->trainIdx].pt));
            obj_project.push_back(obj_keypoint[m->trainIdx].pt);
            img_points.push_back(img_keypoint[m->queryIdx].pt);
        }

        // Determine whether each matched feature is an inlier or not
        std::vector<int> inlier;
        cv::solvePnPRansac(obj_points, img_points, K, dist_coeff, rvec, tvec, false, 500, 2, 0.99, inlier);
        cv::Mat inlier_mask = cv::Mat::zeros(int(match.size()), 1, CV_8U);
        for (size_t i = 0; i < inlier.size(); i++) inlier_mask.at<uchar>(inlier[i]) = 1;
        cv::Mat image_result;
        cv::drawMatches(image, img_keypoint, obj_image, obj_keypoint, match, image_result, cv::Vec3b(0, 0, 255), cv::Vec3b(0, 127, 0), inlier_mask);

        // Estimate camera pose with inliers
        size_t inlier_num = inlier.size();
        if (inlier_num > min_inlier_num)
        {
            std::vector<cv::Point3f> obj_inlier;
            std::vector<cv::Point2f> img_inlier;
            for (int idx = 0; idx < inlier_mask.rows; idx++)
            {
                if (inlier_mask.at<uchar>(idx))
                {
                    obj_inlier.push_back(obj_points[idx]);
                    img_inlier.push_back(img_points[idx]);
                }
            }
            cv::solvePnP(obj_points, img_points, K, dist_coeff, rvec, tvec);

            // Draw the box on the image
            cv::Mat line_lower, line_upper;
            cv::projectPoints(box_lower, rvec, tvec, K, dist_coeff, line_lower);
            cv::projectPoints(box_upper, rvec, tvec, K, dist_coeff, line_upper);
            line_lower.reshape(1).convertTo(line_lower, CV_32S); // Change 4 x 1 matrix (CV_64FC2) to 4 x 2 matrix (CV_32SC1)
            line_upper.reshape(1).convertTo(line_upper, CV_32S); //  because 'cv::polylines()' only accepts 'CV_32S' depth.
            cv::polylines(image_result, line_lower, true, cv::Vec3b(255, 0, 0), 2);
            for (int i = 0; i < line_lower.rows; i++)
                cv::line(image_result, cv::Point(line_lower.row(i)), cv::Point(line_upper.row(i)), cv::Vec3b(0, 255, 0), 2);
            cv::polylines(image_result, line_upper, true, cv::Vec3b(0, 0, 255), 2);
        }

        // Show the image
        cv::String info = cv::format("Inliers: %d (%d%%), Focal Length: %.0f", inlier_num, 100 * inlier_num / match.size(), K.at<double>(0));
        cv::putText(image_result, info, cv::Point(5, 15), cv::FONT_HERSHEY_PLAIN, 1, cv::Vec3b(0, 255, 0));
        cv::imshow("3DV Tutorial: Pose Estimation (Book)", image_result);
        int key = cv::waitKey(1);
        if (key == 27) break; // 'ESC' key: Exit
    }

    video.release();
    return 0;
}

File: scripts/pose_estimation_book1.py
import cv2
import numpy as np

def main():
    input, cover = "../bin/data/blais.mp4", "../bin/data/blais.jpg"
    f, cx, cy = 1000, 320, 240
    min_inlier_num = 100

    # Load the object image and extract features
    obj_image = cv2.imread(cover)

    fdetector = cv2.ORB_create()
    fmatcher = cv2.DescriptorMatcher_create("BruteForce-Hamming")

    obj_keypoint, obj_descriptor = fdetector.detectAndCompute(obj_image, None)
    if (len(obj_keypoint)==0 or len(obj_descriptor)==0): raise Exception("No orb keypoints")
    # obj_descriptor = fmatcher.add(obj_descriptor)

    # Open a video
    cap = cv2.VideoCapture(input)

    # Prepare a box for simple AR
    box_lower = np.array([[30, 145, 0], [30, 200, 0], [200, 200, 0], [200, 145, 0]], dtype=np.float32)
    box_upper = np.array([[30, 145, -50], [30, 200, -50], [200, 200, -50], [200, 145, -50]], dtype = np.float32)

    # Run pose extimation
    K = np.array([[f, 0, cx], [0, f, cy], [0, 0, 1]], dtype=np.float32)
    dist_coeff = np.zeros(5)
    while True:
        # Grab  an image from the video
        ret, image = cap.read()
        if not ret: break

        # Extract features and match them to the object features
        img_keypoint, img_descriptor = fdetector.detectAndCompute(image, None)
        if (len(img_keypoint)==0 or len(img_descriptor)==0): continue


        match = fmatcher.match(img_descriptor, obj_descriptor)
        if len(match) < min_inlier_num: continue

        obj_points, obj_project, img_points  = np.zeros((len(match), 3)), [], []
        for i, m in enumerate(match):
            obj_points[i, :2] = obj_keypoint[m.trainIdx].pt
            obj_project.append(obj_keypoint[m.trainIdx].pt)
            img_points.append(img_keypoint[m.queryIdx].pt)

        obj_points = np.array(obj_points, dtype=np.float32)
        obj_project = np.array(obj_project, dtype=np.float32)
        img_points = np.array(img_points, dtype=np.float32)

        # Deterimine whether each matched feature is an inlier or not
        ret, rvec, tvec, inlier = cv2.solvePnPRansac(objectPoints=obj_points, 
                                                    imagePoints=img_points, 
                                                    cameraMatrix=K, 
                                                    distCoeffs=dist_coeff, 
                                                    useExtrinsicGuess=False, 
                                                    iterationsCount=500, 
                                                    reprojectionError=2., 
                                                    confidence=0.99)
        
        # print("Match data is ", type(match), len(match))
        # inlier_mask = np.zeros(len(match))


        draw_params = dict(matchColor = (0,255,0), # draw matches in green color
                   singlePointColor = (0,127,0),
                   matchesMask = inlier, #inlier_mask, # draw only inliers
                   flags = 2)        
        image_result = cv2.drawMatches(image, img_keypoint, obj_image, obj_keypoint, match, None, **draw_params)

        # Estimate camera pose with inliers
        try:
            inlier_num = len(inlier)
        except:
            inlier_num = 0

        if inlier_num > min_inlier_num:
            ret, rvec, tvec = cv2.solvePnP(obj_points, img_points, K, dist_coeff)

            # Draw the box on the image
            line_lower, _ = cv2.projectPoints(box_lower, rvec, tvec, K, dist_coeff)
            line_upper, _ = cv2.projectPoints(box_upper, rvec, tvec, K, dist_coeff)

            image_result = cv2.polylines(image_result, np.int32([line_lower]), True, (255, 0, 0), 2)
            image_result = cv2.polylines(image_result, np.int32([line_upper]), True, (0, 0, 255), 2)
            for i in range(len(line_lower)):
                image_result = cv2.line(image_result, tuple(line_lower[i][0]), tuple(line_upper[i][0]), (0, 255, 0), 2, cv2.LINE_AA)


        # Show the image
        info = f"Inliers: {inlier_num*100/len(match):.3f} , Focal length: {f}"
        image_result = cv2.putText(image_result, info, (5, 15), cv2.FONT_HERSHEY_PLAIN, 1, (0, 255, 0), 2)
        if cv2.waitKey(1) == ord('q'): break
        cv2.imshow("3DV Tutorial: Pose Estimation (Book)", image_result)

    cap.release()

if __name__ == "__main__":
    main()